sphinxext/opengraph/visitor.py

```diff
@@ -0,0 +1,63 @@
+from docutils import nodes
+
+
+class OpenGraphVisitor(nodes.GenericNodeVisitor):
+    def __init__(
+        self, document: nodes.document, desc_len: int, title_count: int = 1
+    ) -> None:
+        super().__init__(document)
+        self.description = ""
+        self.desc_len = desc_len
+        self.title_count = title_count
+        self.first_list = ""  # todo: rename
+
+    def default_visit(self, node: nodes.Element) -> None:
+        if len(self.description) >= self.desc_len:
+            # Stop the traversal if the description is long enough
+            # raise nodes.StopTraversal
+            pass
+
+        if isinstance(node, (nodes.Invisible, nodes.Admonition)):
+            # Skip all comments and admonitions
+            raise nodes.SkipNode
+
+        if isinstance(node, nodes.title):
+            # title count refers to the amount of titles remaining
+            self.title_count -= 1
+            if self.title_count < 0:
```

Comment:
`count`s should count upwards.

Also, what if the first section does not have text?
Example: https://docs.wpilib.org/en/latest/docs/software/driverstation/manually-setting-the-driver-station-to-start-custom-dashboard.html

If we start counting sections, we should either:
1. Start counting at the first section containing text
2. Only count sections containing text.

Reply:
I forgot to ask about this part in the discord, but this is another part I wasn't really sure about.
Right now the idea is the user can enter how many sections to search, but another possibility would be to only grab the first section that isn't empty.
I agree that regardless empty sections should be ignored.
Not sure if we should put in the titles of empty sections or ignore all titles as it is now.

Reply:
Regarding titles, the options I see are either:
1. Only include titles of sections with text
2. Ignore all titles.

Either is fine really.

Regarding counting sections:
Counting sections falls counter to description length. If the user specifies 300 characters, then it may not be intuitive as to why a lot less characters were included.

I'm fine with providing options for both description_length and num_sections but it would need to be clear in the readme that the amount of text grabbed is `pseudo_min(description_length, num_sections)`.

Reply:
Regarding the counting sections, my train of thought is that `ogp:description` is meant to be a small short description of the page, just grabbing the first X chars means it will always just be everything that can fit, while grabbing the first paragraph will most likely be some kind of either description of the page, or snippet of the contents of the page.
I wouldn't be opposed to having it only grab the first section with text and renaming the `desc_len` to `max_desc_len`.
The other option would be to have that configurable, where it can grab the first 300 chars and ignore all titles, or do as mentioned above.
We could get into more complex title impls, but I feel like the thing I said above should be the default and be the best in most use cases.

Reply:
Grabbing just one section isn't enough. The previous impl by a Sphinx Maintainer was along the lines of `pseudomin(3 sections, 200 char)`. I agree with your argument for counting sections. I just think that both should be options.